### ISSUE TITLE
PHP7

### ISSUE THREAD

**Issue by DRvanR** [simplesamlphp/saml2#91]
Seeing that PHP5.6 has reached the [end of it's active support period][1] as of January 19th 2017 and PHP7.1 is considered the current stable version, I would like to propose a schedule for starting support of features only available in PHP7.

As a timeline, I propose the following:

As of July 19th 2017 (6 months after active support for PHP5.6 ended) we no longer actively support PHP 5.6 in development and create a PHP5.6 compatible branch.
After that, we allow the introduction of PHP7 only features (e.g. return type declarations and scalar typehints) in the master development and thus effectively no longer actively support PHP5.6. This also means a new major should be released.
We should support the 5.6 compatible branch until December 31st 2018 (end of security updates period for PHP 5.6) with bugfixes and security updates _only_ .

This means three things:

1. Active work on PHP7 support can be started on a non-master branch
2. There is a clear Supported PHP Versions timeline (which can be communicated)
3. There still are 6 months to ensure that any code can work with PHP7.

As to the limited support (bugfixes and security updates only), this is a deliberate choice - it should not be encouraged to use unsupported and possibly insecure versions of PHP.

Thoughts?

[1]: http://php.net/supported-versions.php

**Comment by thijskh** [comment]
My thought is that what the PHP website claims about the supported versions of PHP is not necessarily a complete picture of supported versions. Hardly anyone downloads and compiles PHP from the PHP website these days; they use distribution packages which have different (and often longer) support timelines.

**Comment by jaimeperez** [comment]
I'm with Thijs on this.

The fact that PHP drops support for a specific version doesn't mean that version is not supported via other means. As an example, I believe Red Hat is still backporting security updates to PHP 5.3.

This is a bit of a chicken & egg problem. We can't really drop support for very old versions of PHP given that those versions are still supported by some major vendors, and at the same time people won't update their PHP installations because things keep working.

In any case, I agree with starting the work on a PHP7 branch and taking advantage of it, but I would like to still backport enhancements to older branches.

**Comment by DRvanR** [comment]
Since this is a library, not an application, and it is platform (OS) independent, I'm of the opinion that we should not try to support as much different versions of PHP as possible because some platform(s) still offer that version and backport security updates. An important reason for this is that there are language features that are good to use that are simply not possible in older releases. Also, we'll have to deal with potentially different API's in different versions of the library (think return type declarations, or scalar typehints). This has in part to do with significant changes in the underlying engine, for instance the introduction of the AST in PHP7, and the current work on a JIT compiler for PHPnextmajor. Also, there are significant changes in the logic, implementation and language constructs between PHP5 and PHP7, which sometimes are also BC breaks. This makes trying to support as much as we can tricky and to me only adds overhead. The maintenance overhead with backports between different major versions can be significant. 

W.r.t. chicken and egg: all current supported versions have a working version. Under the proposed timeline, bugfixes and security updates will be backported - we could even choose to extend that period. Therefor, each version from 5.3 onward has a working version. They simply will not receive new functionality after 6 months from now - ample time to consider the impact I'd say :wink: Only new features & functional changes will not be backported, which makes making these changes simpler. Trying to backport when the versions offer different API's might prove difficult.

I do realize this is a bit a philosophical discussion however where the arguments will stem from preference of being the chicken or the egg :smile: - I'm not adamant we should not support older versions, I'm only wary of doing so. Also, I'd like to focus the effort on improving the library rather than spend time making ports to different versions. Hence the proposal.

As a simple alternative: just do not use PHP7-only features. This might prove a worse solution in the long run when having to upgrade because we must though - or not, only time will tell.

Main point is: a decision should be made, and that decision should be communicated.

**Comment by tvdijen** [comment]
> I believe Red Hat is still backporting security updates to PHP 5.3.
Correct, the default PHP version on RHEL6 is 5.3 and for RHEL7 it's 5.4.

**Comment by jaimeperez** [comment]
Daan, I honestly can't argue against any of your arguments. In fact, I agree with most of them.

However, I have to wear also my fedop hat here, and as such, I really need SSP (and by extension, the SAML2 library) to keep supporting older versions of PHP. We don't run our own infrastructure (Univ. of Oslo does that for us) and for now we are stuck with the version of PHP shipped with RHEL. If we have such limitation, considering SSP is key for us and that we had over 150 mill. logins last year (my point being: we have a big infrastructure), I could totally imagine most of our users to be stuck in the same way or even worse with legacy systems that are still maintained in one way or another.

I could just say "let's freeze the version of the SAML2 library we are using in SSP and stick to that", but at the same time it wouldn't surprise me if we really need new features from it in the near future (at least not if I look at the recent past), so I'm a bit hesitant. On the other hand, as I said, I agree with you and at some point we really need to make the move to PHP 7 only.

@thijskh, @tvdijen, a penny for your thoughts. How would you go forward?

**Comment by tvdijen** [comment]
Well, looking at our situation here, PHP 7.x is simply not available...  5.6 would be possible, since it is provided by RedHat through the EPEL-repository.
PHP7 imho is one bridge too far. If your users stick to an old (frozen) version, you'll eventually lose them as a whole. They're not gonna come up with PR's for a version they can't use themselves.
To me it would make sense to drop support for 5.3-5.5 in the next major release.

**Comment by pmeulen** [comment]
My 2c:

It is a given that at some point in the future we will start using PHP 7 only functionality in the library, or stop testing with PHP 5.x and thus break PHP 5.x compatibility. The question is when.

We shouldn't drop support for PHP 5 anytime soon. There will be users using the library on e.g. Debian/Ubuntu LTS, Redhat for several years to come. This means that we need supply security and bugfixes to the PHP 5 version of the library. I don't see anyone disagreeing with that. The question is what new features we want to make available to the PHP 5 version of the library.

I expect distributions to switch tot PHP 7 in their next iteration. Debian made such an announcement [1] and allows multiple versions of PHP to be installed at the same time. RedHat has their software collections allowing the same. With php-fpm and composer for package management being mature users aren't as locked in to running everything on the same conservative distro provided PHP version as they were before. At some point we will want to move to PHP7 because the libraries that we depend on have moved as well.

If we drop support for PHP < 7 now think we're on the early side. Very few of our users will be using PHP 7, thus what we write will not be usable to them, and will not get testing. The library is already compatible with PHP 7, we keep it that way obviously.

The question is what are the extra costs for us of supporting the older PHP versions for new features? Do we have anything planned where we expect to run into issues? Is there a middle ground, will it help to stop targeting PHP < 5.6 for instance?

[1] https://lists.debian.org/debian-devel-announce/2016/01/msg00002.html

**Comment by DRvanR** [comment]
> The question is what are the extra costs for us of supporting the older PHP versions for new features? Do we have anything planned where we expect to run into issues? Is there a middle ground, will it help to stop targeting PHP < 5.6 for instance?

Frankly: unknown. Unsure as to what features, changes and fixes might be wanted/needed so can't really say. There are however quite some BC breaks in between 5.3 and 7.1 that might influence us - Feide still uses 5.3 and cannot (?) migrate to a newer version. The build still works in 5.3-7.0 though, so for now it should be fine. It mostly limits what features to use and care should be taken not to introduce features/changes that might break in some version, which simply adds development overhead. At the same time very old versions of libraries are used for development dependencies (e.g. PHPunit 3.x range is not even documented anymore, the next major release 4.0 is 3 years old) which might prove troublesome to use as the APIs are unfamiliar.

To me the biggest concerns (and the reason why I opened this issue with a deliberately ambitious proposal :wink: ) to me are:

- incompatible code (see #90 for an example) 
- the issue of other dependencies in a project that uses multiple libraries dropping support for php <7
- the issue of libraries we depend on dropping support for php <7

The first issue is a real issue, as can be seen in the PR. The second issue is maybe not that serious, and as long as it works on PHP7 it should be fine. With respect to the last issue, that does not have to be an issue. I don't expect the xmlseclibs to be updated any time soon :wink: (for development dependencies see above)

---

As the consensus seems to be that the proposed timeline is too quick, anyone willing to provide an alternative? For reference, the proposed timeline was:

1. As of July 19th 2017 (6 months after active support for PHP5.6 ended) we no longer actively support PHP 5.6 in development and create a PHP5.6 compatible branch. After that, we allow the introduction of PHP7 only features (e.g. return type declarations and scalar typehints) in the master development and thus effectively no longer actively support PHP5.6 in development. This also means a new major should be released.
2. We should support the 5.6 compatible branch until December 31st 2018 (end of security updates period for PHP 5.6) with bugfixes and security updates _only_.

The assumption was made that nobody would still be running PHP <5.6 - which turns out to be faulty, as most assumptions are :wink:

Thoughts, or rather, proposals?

**Comment by tvdijen** [comment]
We've decided to start working on a 4.0 branch where we drop support for PHP 5.x and start taking advantage of the benefits that PHP7 has to offer (especially type hinting).
We've branched of a 3.x branch from master that will receive bugfixes for a while and some work on a 4.0 release is already done and awaiting approval.
